symforce/slam/imu_preintegration/imu_factor.h

```diff
@@ -0,0 +1,192 @@
+ *
+ * Example Usage:
+ *
+ *   enum Var : char {
+ *     POSE = 'p',        // Pose3d
+ *     VELOCITY = 'v',    // Vector3d
+ *     ACCEL_BIAS = 'A',  // Vector3d
+ *     GYRO_BIAS = 'G',   // Vector3d
+ *     GRAVITY = 'g',     // Vector3d
+ *     EPSILON = 'e'      // Scalar
+ *   };
+ *
+ *   struct ImuMeasurement {
+ *     Eigen::Vector3d acceleration;
+ *     Eigen::Vector3d angular_velocity;
+ *     double duration;
+ *   };
+ *
+ *   int main() {
+ *     // Dummy function declarations
+ *     std::vector<ImuMeasurement> GetMeasurementsBetween(double start_time, double end_time);
+ *     std::vector<double> GetKeyFrameTimes();
+ *     Eigen::Vector3d GetAccelBiasEstimate(double time);
+ *     Eigen::Vector3d GetGyroBiasEstimate(double time);
+ *     void AppendOtherFactors(std::vector<sym::Factord> & factors);
+ *
+ *     // Example accelerometer and gyroscope covariances
+ *     const Eigen::Matrix3d accel_cov = 1e-5 * Eigen::Matrix3d::Identity();
+ *     const Eigen::Matrix3d gyro_cov = 1e-5 * Eigen::Matrix3d::Identity();
+ *
+ *     std::vector<sym::Factord> factors;
+ *
+ *     // GetKeyFrameTimes assumed to return at least 2 times
+ *     std::vector<double> key_frame_times = GetKeyFrameTimes();
+ *
+ *     // Integrating Imu measurements between keyframes, creating an ImuFactor for each
+ *     for (unsigned int i = 0; i < key_frame_times.size() - 1; i++) {
+ *       const double start_time = key_frame_times[i];
+ *       const double end_time = key_frame_times[i + 1];
+ *
+ *       const std::vector<ImuMeasurement> measurements = GetMeasurementsBetween(start_time,
+ *                                                                               end_time);
+ *
+ *       // ImuPreintegrator should be initialized with the best guesses for the IMU biases
+ *       sym::ImuPreintegrator<double> integrator(GetAccelBiasEstimate(start_time),
+ *                                                GetGyroBiasEstimate(start_time));
+ *       for (const ImuMeasurement& meas : measurements) {
+ *         // IntegrateMeasurement assumes IMU measures acceleration of magnitude |gravity| pointing
+ *         // away from the Earth when stationary
+ *         integrator.IntegrateMeasurement(meas.acceleration, meas.acceleration, accel_cov,
+ *                                         gyro_cov, meas.duration);
+ *       }
+ *
+ *       factors.push_back(sym::ImuFactor<double>(integrator)
+ *                             .Factor({{Var::POSE, i},
+ *                                      {Var::VELOCITY, i},
+ *                                      {Var::POSE, i + 1},
+ *                                      {Var::VELOCITY, i + 1},
+ *                                      {Var::ACCEL_BIAS, i},
+ *                                      {Var::GYRO_BIAS, i},
+ *                                      {Var::GRAVITY},
+ *                                      {Var::EPSILON}}));
+ *     }
+ *
+ *     AppendOtherFactors(factors);
```

Comment:
I think we should probably leave this out or have a comment like `// Add any other factors for the optimization problem` or something

Reply:
Hmm. The reason I wanted to add other factors is because, as this example stands, since we have a different bias for every factor, the problem is under-constrained, and I thought it would be good to draw attention to the fact that the `ImuFactor` is meant to be used with other factors (unless you already have a good idea of all the biases and just choose to not optimize them) (related, but since we have no priors that also leads to it being under-constrained in another way).

Then, my worry about simply adding a comment is it assumes the reader already knows how to add any factors (which might not be immediately clear if they are still writing their first problem). I thought by saying `AppendOtherFacotrs(factors)`, it would be more demonstrative by showing that you add other factors by appending to the same vector.

Reply:
Yep, I was thinking it's ok to leave it implied / assumed that you need to fix things or have other measurements for things to be constrained, but that makes sense - adding a comment would just be to make it more clear that these are "any other factors you might have for the problem" and not some specific set of factors required for the ImuFactor

Reply:
Sounds fine. Since I didn't think it would hurt much relative to the added clarity, I added the comment right above saying
 `// Adding any other factors there might be for the problem`